File: sdk/flutter/iapstack/test/client_test.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:iapstack/iapstack.dart';
import 'package:test/test.dart';

void main() {
  group('IapStackClient', () {
    test('sends the exact verify contract and decodes projections', () async {
      late http.Request captured;
      final client = IapStackClient(
        _config(),
        httpClient: MockClient((request) async {
          captured = request;
          return http.Response(jsonEncode(_verificationJson), 200);
        }),
      );

      final result = await client.verifyPurchase(_purchase(),
          requestId: 'request-client-1');

      expect(captured.method, 'POST');
      expect(captured.url.path,
          '/proxy/v1/applications/application-1/purchases:verify');
      expect(captured.headers['Authorization'], 'Bearer customer-token');
      expect(captured.headers['Content-Type'], 'application/json');
      expect(captured.headers['X-Request-ID'], 'request-client-1');
      expect(captured.headers['X-IAPStack-SDK'], startsWith('flutter/'));
      expect(jsonDecode(captured.body), <String, Object?>{
        'external_customer_id': 'customer-external',
        'claimed_products': <String>['premium_lifetime'],
        'evidence': <String, Object?>{
          'purchase_data': '{"productId":"premium_lifetime"}',
          'signature': 'signed',
          'product_kind': 'non_consumable',
        },
      });
      expect(result.customerId, 'customer-internal');
      expect(result.verifiedAt, DateTime.utc(2026, 8, 24, 20));
      expect(result.entitlements.single.grantsAccess, isTrue);
      expect(result.entitlements.single.effectiveEndsAt, isNull);
    });

    test('sends restore batches and preserves result order', () async {
      late Object? body;
      final client = IapStackClient(
        _config(),
        httpClient: MockClient((request) async {
          body = jsonDecode(request.body);
          return http.Response(
            jsonEncode(<String, Object?>{
              'results': <Object?>[_verificationJson, _verificationJson]
            }),
            200,
          );
        }),
      );

      final result = await client.restorePurchases(
          <PurchaseSubmission>[_purchase(), _purchase('pro_monthly')]);

      final json = body! as Map<String, dynamic>;
      expect((json['purchases']! as List<dynamic>), hasLength(2));
      expect(result.results, hasLength(2));
    });

    test('fails closed when an allowed entitlement reaches its effective end',
        () {
      final endsAt = DateTime.utc(2026, 8, 26, 12);
      final entitlement = Entitlement(
        key: 'premium',
        access: 'allowed',
        reason: 'canceled_at_period_end',
        version: 7,
        effectiveStartsAt: endsAt.subtract(const Duration(days: 30)),
        effectiveEndsAt: endsAt,
      );

      expect(
          entitlement
              .grantsAccessAt(endsAt.subtract(const Duration(microseconds: 1))),
          isTrue);
      expect(entitlement.grantsAccessAt(endsAt), isFalse);
      expect(entitlement.grantsAccessAt(endsAt.add(const Duration(hours: 1))),
          isFalse);
    });

    test('escapes customer identifiers as one path segment', () async {
      late Uri captured;
      final client = IapStackClient(
        _config(),
        httpClient: MockClient((request) async {
          captured = request.url;
          return http.Response(
            jsonEncode(<String, Object?>{
              'customer_id': 'customer-internal',
              'entitlements': _verificationJson['entitlements'],
            }),
            200,
          );
        }),
      );

      await client.getEntitlements('customer/with space');

      expect(captured.pathSegments, contains('customer/with space'));
      expect(captured.toString(), contains('customer%2Fwith%20space'));
    });

    test('retries transient responses and returns the eventual result',
        () async {
      var attempts = 0;
      final client = IapStackClient(
        _config(
          retryPolicy: const IapStackRetryPolicy(
            maxAttempts: 2,
            baseDelay: Duration.zero,
            maxDelay: Duration.zero,
          ),
        ),
        httpClient: MockClient((request) async {
          attempts++;
          if (attempts == 1) {
            return http.Response(
              jsonEncode(<String, Object?>{
                'error': <String, Object?>{
                  'code': 'provider_unavailable',
                  'message': 'provider is unavailable',
                  'request_id': 'request-server-1',
                },
              }),
              503,
            );
          }
          return http.Response(jsonEncode(_verificationJson), 200);
        }),
      );

      final result = await client.verifyPurchase(_purchase());

      expect(attempts, 2);
      expect(result.customerId, 'customer-internal');
    });

    test('exposes stable API errors without response payloads', () async {
      final client = IapStackClient(
        _config(retryPolicy: const IapStackRetryPolicy(maxAttempts: 1)),
        httpClient: MockClient((request) async => http.Response(
              jsonEncode(<String, Object?>{
                'error': <String, Object?>{
                  'code': 'invalid_purchase',
                  'message': 'purchase evidence could not be verified',
                  'request_id': 'request-server-2',
                },
                'secret': 'must-not-escape',
              }),
              422,
            )),
      );

      await expectLater(
        client.verifyPurchase(_purchase()),
        throwsA(
          isA<IapStackApiException>()
              .having((error) => error.statusCode, 'statusCode', 422)
              .having((error) => error.code, 'code', 'invalid_purchase')
              .having(
                  (error) => error.requestId, 'requestId', 'request-server-2')
              .having((error) => error.retryable, 'retryable', isFalse)
              .having((error) => error.toString(), 'safe string',
                  isNot(contains('must-not-escape'))),
        ),
      );
    });

    test('rejects oversized and invalid JSON responses', () async {
      final oversized = IapStackClient(
        _config(maxResponseBytes: 8),
        httpClient: MockClient(
            (request) async => http.Response('{"value":"too-large"}', 200)),
      );
      final invalid = IapStackClient(
        _config(),
        httpClient: MockClient(
            (request) async => http.Response('<html>proxy error</html>', 200)),
      );

      await expectLater(oversized.verifyPurchase(_purchase()),
          throwsA(isA<IapStackProtocolException>()));
      await expectLater(invalid.verifyPurchase(_purchase()),
          throwsA(isA<IapStackProtocolException>()));
    });

    test('turns a bounded attempt deadline into a timeout exception', () async {
      final client = IapStackClient(
        _config(
          timeout: const Duration(milliseconds: 1),
          retryPolicy: const IapStackRetryPolicy(maxAttempts: 1),
        ),
        httpClient: MockClient((request) async {
          await Future<void>.delayed(const Duration(milliseconds: 20));
          return http.Response(jsonEncode(_verificationJson), 200);
        }),
      );

      await expectLater(client.verifyPurchase(_purchase()),
          throwsA(isA<IapStackTimeoutException>()));
    });

    test('aborts the underlying HTTP request when an attempt times out',
        () async {
      final transport = _AbortTrackingClient();
      final client = IapStackClient(
        _config(
          timeout: const Duration(milliseconds: 1),
          retryPolicy: const IapStackRetryPolicy(maxAttempts: 1),
        ),
        httpClient: transport,
      );

      await expectLater(client.verifyPurchase(_purchase()),
          throwsA(isA<IapStackTimeoutException>()));
      await transport.aborted.timeout(const Duration(seconds: 1));
    });
  });

  group('IapStackRetryPolicy', () {
    test('applies full jitter within the exponential cap', () {
      const policy = IapStackRetryPolicy(
        baseDelay: Duration(milliseconds: 100),
        maxDelay: Duration(milliseconds: 250),
      );

      expect(policy.delayAfter(1, randomValue: 0), Duration.zero);
      expect(policy.delayAfter(1, randomValue: 0.5),
          const Duration(milliseconds: 50));
      expect(policy.delayAfter(4, randomValue: 1),
          const Duration(milliseconds: 250));
      expect(() => policy.delayAfter(1, randomValue: 1.1), throwsArgumentError);
    });
  });

  group('IapStackConfig', () {
    test('requires HTTPS unless insecure HTTP is explicit', () {
      expect(
        () => IapStackClient(
          IapStackConfig(
            baseUri: Uri.parse('http://iap.example'),
            applicationId: 'application-1',
            customerToken: 'customer-token',
          ),
          httpClient: MockClient((request) async => http.Response('{}', 200)),
        ),
        throwsArgumentError,
      );
    });

    test('redacts invalid bearer values from validation errors', () {
      expect(
        () => IapStackClient(
          IapStackConfig(
            baseUri: Uri.parse('https://iap.example'),
            applicationId: 'application-1',
            customerToken: 'secret with spaces',
          ),
          httpClient: MockClient((request) async => http.Response('{}', 200)),
        ),
        throwsA(isA<ArgumentError>().having((error) => error.toString(),
            'message', isNot(contains('secret with spaces')))),
      );
    });
  });
}

IapStackConfig _config({
  Duration timeout = const Duration(seconds: 1),
  IapStackRetryPolicy retryPolicy = const IapStackRetryPolicy(maxAttempts: 1),
  int maxResponseBytes = 1024 * 1024,
}) =>
    IapStackConfig(
      baseUri: Uri.parse('https://iap.example/proxy'),
      applicationId: 'application-1',
      customerToken: 'customer-token',
      timeout: timeout,
      retryPolicy: retryPolicy,
      maxResponseBytes: maxResponseBytes,
    );

PurchaseSubmission _purchase([String productId = 'premium_lifetime']) =>
    PurchaseSubmission(
      externalCustomerId: 'customer-external',
      claimedProducts: <String>[productId],
      evidence: <String, Object?>{
        'purchase_data': '{"productId":"$productId"}',
        'signature': 'signed',
        'product_kind': 'non_consumable',
      },
    );

final Map<String, Object?> _verificationJson = <String, Object?>{
  'verified_at': '2026-08-24T20:00:00Z',
  'customer_id': 'customer-internal',
  'entitlements': <Object?>[
    <String, Object?>{
      'key': 'premium',
      'access': 'allowed',
      'reason': 'purchase_valid',
      'effective_starts_at': '2026-08-24T19:00:00Z',
      'version': 1,
    },
  ],
};

final class _AbortTrackingClient extends http.BaseClient {
  final Completer<void> _aborted = Completer<void>();

  Future<void> get aborted => _aborted.future;

  @override
  Future<http.StreamedResponse> send(http.BaseRequest request) async {
    if (request is! http.Abortable || request.abortTrigger == null) {
      throw StateError('request was not abortable');
    }
    await request.abortTrigger;
    if (!_aborted.isCompleted) {
      _aborted.complete();
    }
    throw http.RequestAbortedException(request.url);
  }
}

File: sdk/react-native/test/client.test.ts
import { test } from 'bun:test';
import assert from 'node:assert/strict';
import {
  appleEvidence,
  Entitlement,
  googlePlayEvidence,
  huaweiEvidence,
  IapStackApiError,
  IapStackClient,
  IapStackConfig,
  IapStackProtocolError,
  IapStackRetryPolicy,
  IapStackTimeoutError,
} from '../src/index';

type MockRequest = {
  url: string;
  init: RequestInit;
};

function withMockFetch(
  handler: (input: RequestInfo | URL, init: RequestInit) => Promise<Response>,
): { calls: MockRequest[]; restore: () => void } {
  const originalFetch = globalThis.fetch;
  const calls: MockRequest[] = [];
  globalThis.fetch = async (input: RequestInfo | URL, init: RequestInit = {}) => {
    calls.push({
      url: input instanceof URL ? input.toString() : `${input}`,
      init,
    });
    return handler(input, init);
  };

  return {
    calls,
    restore() {
      globalThis.fetch = originalFetch;
    },
  };
}

function verificationPayload() {
  return {
    verified_at: '2026-09-09T00:00:00.000Z',
    customer_id: 'customer-123',
    entitlements: [
      {
        key: 'premium',
        access: 'allowed',
        reason: 'purchase_valid',
        version: 1,
        effective_starts_at: '2026-09-09T00:00:00.000Z',
        effective_ends_at: null,
      },
    ],
  };
}

function restorePayload() {
  return {
    results: [
      {
        verified_at: '2026-09-09T00:00:00.000Z',
        customer_id: 'customer-123',
        entitlements: [],
      },
    ],
  };
}

function clientConfig(overrides: Partial<ConstructorParameters<typeof IapStackConfig>[0]> = {}) {
  return new IapStackConfig({
    baseUri: 'https://iapstack.test/api',
    applicationId: 'app-1',
    customerToken: 'token',
    ...overrides,
  });
}

const samplePurchase = {
  externalCustomerId: 'customer-123',
  claimedProducts: ['premium_annual'],
  evidence: googlePlayEvidence({
    purchaseToken: 'token-abc',
    productKind: 'subscription',
  }),
};

test('verifyPurchase sends correct method, path, headers, and payload', async () => {
  const { calls, restore } = withMockFetch(async (_input, _init) => {
    return new Response(JSON.stringify(verificationPayload()), {
      status: 200,
      headers: { 'content-type': 'application/json' },
    });
  });

  try {
    const client = new IapStackClient(clientConfig());
    const result = await client.verifyPurchase(samplePurchase, 'request-123');

    assert.equal(result.customerId, 'customer-123');
    assert.equal(result.entitlements[0]?.grantsAccess, true);
    assert.equal(calls.length, 1);
    assert.equal(
      calls[0].url,
      'https://iapstack.test/api/v1/applications/app-1/purchases:verify',
    );

    const headers = new Headers(calls[0].init.headers as HeadersInit);
    assert.equal(headers.get('Accept'), 'application/json');
    assert.equal(headers.get('Authorization'), 'Bearer token');
    assert.equal(headers.get('X-IAPStack-SDK'), 'react-native/0.1.0-dev.1');
    assert.equal(headers.get('X-Request-ID'), 'request-123');
    assert.equal(headers.get('Content-Type'), 'application/json');

    const requestBody = JSON.parse((calls[0].init.body as string) || '{}');
    assert.equal(requestBody.external_customer_id, 'customer-123');
    assert.deepEqual(requestBody.claimed_products, ['premium_annual']);
    assert.equal(requestBody.evidence.purchase_token, 'token-abc');
  } finally {
    restore();
  }
});

test('restorePurchases retries only on retryable status codes', async () => {
  let attempt = 0;
  const { calls, restore } = withMockFetch(async (_input, _init) => {
    attempt += 1;
    if (attempt === 1) {
      return new Response(
        JSON.stringify({
          error: {
            code: 'service_unavailable',
            message: 'temporary',
          },
        }),
        {
          status: 503,
          headers: {
            'content-type': 'application/json',
            'x-request-id': 'req-503',
          },
        },
      );
    }
    return new Response(JSON.stringify(restorePayload()), {
      status: 200,
      headers: {
        'content-type': 'application/json',
        'x-request-id': 'req-ok',
      },
    });
  });

  try {
    const client = new IapStackClient(
      clientConfig({
        retryPolicy: new IapStackRetryPolicy({
          maxAttempts: 2,
          baseDelayMs: 0,
          maxDelayMs: 0,
        }),
      }),
    );
    const result = await client.restorePurchases([samplePurchase]);

    assert.equal(attempt, 2);
    assert.equal(calls.length, 2);
    assert.equal(
      calls[0].url,
      'https://iapstack.test/api/v1/applications/app-1/purchases:restore',
    );
    assert.equal(calls[1].url, 'https://iapstack.test/api/v1/applications/app-1/purchases:restore');
    assert.equal(result.results[0].customerId, 'customer-123');
  } finally {
    restore();
  }
});

test('non-retryable API errors are surfaced without another attempt', async () => {
  let attempt = 0;
  const { calls, restore } = withMockFetch(async () => {
    attempt += 1;
    return new Response(
      JSON.stringify({
        error: {
          code: 'invalid_request',
          message: 'invalid',
        },
      }),
      {
        status: 400,
        headers: { 'content-type': 'application/json' },
      },
    );
  });

  try {
    const client = new IapStackClient(
      clientConfig({
        retryPolicy: new IapStackRetryPolicy({
          maxAttempts: 3,
          baseDelayMs: 0,
          maxDelayMs: 0,
        }),
      }),
    );

    await assert.rejects(
      () => client.getEntitlements('customer-123'),
      (error) => {
        assert(error instanceof IapStackApiError);
        assert.equal((error as IapStackApiError).statusCode, 400);
        return true;
      },
    );
    assert.equal(attempt, 1);
    assert.equal(calls.length, 1);
  } finally {
    restore();
  }
});

test('request timeouts are converted into IapStackTimeoutError', async () => {
  const { restore } = withMockFetch(async (_input, init) => {
    return new Promise((_resolve, reject) => {
      init.signal?.addEventListener('abort', () => {
        const timeout = new Error('timed out') as Error & { name: string };
        timeout.name = 'AbortError';
        reject(timeout);
      });
    });
  });

  try {
    const client = new IapStackClient(
      clientConfig({
        timeoutMs: 5,
      }),
    );

    await assert.rejects(
      () => client.verifyPurchase(samplePurchase),
      (error) => error instanceof IapStackTimeoutError,
    );
  } finally {
    restore();
  }
});

test('duplicate claimed products are rejected before making requests', async () => {
  const { calls, restore } = withMockFetch(async () => {
    throw new Error('should not be hit');
  });

  try {
    const client = new IapStackClient(clientConfig());
    await assert.rejects(
      () =>
        client.restorePurchases([
          {
            ...samplePurchase,
            claimedProducts: ['same', 'same'],
          },
        ]),
      (error) => {
        assert.match((error as Error).message, /must not contain duplicates/);
        return true;
      },
    );
    assert.equal(calls.length, 0);
  } finally {
    restore();
  }
});

test('empty and non-object evidence is rejected before making requests', async () => {
  const { calls, restore } = withMockFetch(async () => {
    throw new Error('should not be hit');
  });

  try {
    const client = new IapStackClient(clientConfig());
    await assert.rejects(
      () => client.verifyPurchase({ ...samplePurchase, evidence: {} }),
      (error) => {
        assert.match((error as Error).message, /evidence is required/);
        return true;
      },
    );
    await assert.rejects(
      () =>
        client.verifyPurchase({
          ...samplePurchase,
          evidence: [] as unknown as Record<string, unknown>,
        }),
      (error) => {
        assert.match((error as Error).message, /evidence is required/);
        return true;
      },
    );
    assert.equal(calls.length, 0);
  } finally {
    restore();
  }
});

test('rejects oversized responses as protocol errors', async () => {
  const { restore } = withMockFetch(async () => {
    return new Response('{"value":"too-large"}', {
      status: 200,
      headers: { 'content-type': 'application/json' },
    });
  });

  try {
    const client = new IapStackClient(clientConfig({ maxResponseBytes: 8 }));
    await assert.rejects(
      () => client.verifyPurchase(samplePurchase),
      (error) => error instanceof IapStackProtocolError,
    );
  } finally {
    restore();
  }
});

test('aborts when Content-Length exceeds maxResponseBytes', async () => {
  const { restore } = withMockFetch(async () => {
    return new Response('{"ok":true}', {
      status: 200,
      headers: {
        'content-type': 'application/json',
        'content-length': '999999',
      },
    });
  });

  try {
    const client = new IapStackClient(clientConfig({ maxResponseBytes: 32 }));
    await assert.rejects(
      () => client.verifyPurchase(samplePurchase),
      (error) => error instanceof IapStackProtocolError,
    );
  } finally {
    restore();
  }
});

test('wraps contract mismatches as IapStackProtocolError', async () => {
  const { restore } = withMockFetch(async () => {
    return new Response(
      JSON.stringify({
        verified_at: '2026-09-09T00:00:00.000Z',
        customer_id: '',
        entitlements: [],
      }),
      {
        status: 200,
        headers: { 'content-type': 'application/json' },
      },
    );
  });

  try {
    const client = new IapStackClient(clientConfig());
    await assert.rejects(
      () => client.verifyPurchase(samplePurchase),
      (error) => error instanceof IapStackProtocolError,
    );
  } finally {
    restore();
  }
});

test('fails closed when an allowed entitlement reaches its effective end', () => {
  const endsAt = new Date('2026-08-26T12:00:00.000Z');
  const entitlement = new Entitlement({
    key: 'premium',
    access: 'allowed',
    reason: 'canceled_at_period_end',
    version: 7,
    effectiveStartsAt: new Date('2026-07-27T12:00:00.000Z'),
    effectiveEndsAt: endsAt,
  });

  assert.equal(entitlement.grantsAccessAt(new Date(endsAt.getTime() - 1)), true);
  assert.equal(entitlement.grantsAccessAt(endsAt), false);
  assert.equal(entitlement.grantsAccessAt(new Date(endsAt.getTime() + 3600_000)), false);
  assert.equal(
    new Entitlement({
      key: 'premium',
      access: 'denied',
      reason: 'expired',
      version: 1,
    }).grantsAccess,
    false,
  );
});

test('evidence helper builders validate required input', () => {
  const apple = appleEvidence({
    signedTransaction: 'aaa.bbb.ccc',
    productKind: 'subscription',
  });
  assert.equal(apple.product_kind, 'subscription');
  assert.equal(apple.signed_transaction, 'aaa.bbb.ccc');

  const google = googlePlayEvidence({
    purchaseToken: 'play-token',
    productKind: 'non_consumable',
  });
  assert.equal(google.product_kind, 'non_consumable');

  const huawei = huaweiEvidence({
    purchaseData: '{"ok":true}',
    signature: 'sig',
    productKind: 'subscription',
  });
  assert.equal(huawei.signature, 'sig');
  assert.equal(huawei.purchase_data, '{"ok":true}');

  assert.throws(() => {
    googlePlayEvidence({ purchaseToken: 'x' } as never);
  });
  assert.throws(() => {
    appleEvidence({ signedTransaction: 'txn', productKind: 'subscription' });
  });
  assert.throws(() => {
    appleEvidence({ signedTransaction: '   ', productKind: 'subscription' });
  });
  assert.throws(() => {
    huaweiEvidence({
      purchaseData: 'not-json',
      signature: 'sig',
      productKind: 'subscription',
    });
  });
});
